modules/instrument_builder/jsx/redcap2linst.js

```diff
@@ -0,0 +1,92 @@
+      return 'static{@}{@}' + label + '\n';
+  case 'radio':
+  case 'dropdown':
+      // Radio or dropdown maps to a select and the options are in the
+      // same format in the dictionary.
+      return 'select{@}' + redcapfieldname + '{@}' + label + '{@}'
+          + optionsToLINST(redcapChoices) + '\n';
+  case 'checkbox':
+      // checkboxes are the same format as radios but allow multiple options,
+      // so map to a selectmultiple instead of a select
+      return 'selectmultiple{@}' + redcapfieldname + '{@}' + label + '{@}'
+          + optionsToLINST(redcapChoices) + '\n';
+  case 'yesno':
+      // Map yes/no fields to dropdowns with yes and no options.
+      return 'select{@}' + redcapfieldname + '{@}' + label + '{@}'
+          + 'NULL=>\'\'{-}\'yes\'=>\'Yes\'{-}\'no\'=>\'No\'' + '\n';
+  case 'calc':
+      // Calc maps to a score field.
+      // We create the DB field but don't do the score.
+      return 'static{@}' + redcapfieldname + '{@}' + label + '\n';
+  case 'sql':
+      // The "SQL" data type seems to just be for presentation? I hope?
+      return '';
+  case 'slider':
+      return 'numeric{@}' + redcapfieldname + '{@}' + label + '{@}0{@}100'
+          + '\n';
+  case 'file':
+      // File upload - NOT SUPPORTED BY LINST
+      return '';
+  case 'notes':
+      // REDCap calls textareas notes
+      return 'textarea{@}' + redcapfieldname + '{@}' + label + '\n';
+  default:
+      callback.error('syntaxError', 'Unhandled REDCap type ' + redcaptype);
+      return '';
+  }
+};
+
+const optionsToLINST = (dictionary, callback) => {
+  const choices = dictionary.split(' | ');
+  const linstChoices = choices.map((choice, index) => {
+    const matches = choice.match(/^(\s)*(\d+)(\s)*,(.*)$/);
+    if (matches === null) {
+      callback.error('syntaxError', 'Could not parse radio option: ' + choice);
+      return;
+    }
+    const backend = matches[2] + '_' + (matches[4].trim()).replace(/\s+/g, '_');
+    const linstFormat = '\'' + backend.toLowerCase() + '\'=>\''
+      + (matches[4]).trim() + '\'';
+    return linstFormat;
+  });
+
+  return linstChoices.join('{-}');
+};
+
+export const downloadLinst = (inst, linst) => {
+  'use strict';
+  setTimeout(() => {
+    const fileName = inst || 'new_instrument';
+    const element = document.createElement('a');
+    const blob = new Blob([linst.join('')], {type: 'text/plain;base64'});
+    const url = URL.createObjectURL(blob);
+    element.href = url;
+    element.download = fileName + '.linst';
+    element.style.display = 'none';
+    // element.addEventListener('click', () => {
+    //   // remove the element once it has been clicked
+    //   // setTimeout(() => {
+    //   // }, 10);
+    // });
+    // add element to the document so that it can be clicked
+    // this is need to download in firefox
+    document.body.appendChild(element);
+    element.click();
+    document.body.removeChild(element);
+    URL.revokeObjectURL(url);
```

Comment:
hey @maltheism @laemtl can i pick your brain on this race condition?

this bit of code intends to download in the browser some 80+ linst files. I'm running into the issue where revokeObjectURL isn't being called before the next createObjectURL in the loop [here](https://github.com/aces/Loris/pull/8282/files#diff-3db2a0bc8c5cd4dbd2a87129ddbbd17747ffbb8ceae93dd69b004f02ad6bcda3R162), resulting in the browser running out of memory and only 20+ files getting downloaded (i **think** this is what is happening). the most number of files ive been able to download with this code is 60, but most times only 20 or 10 get downloaded.

do you know how I can make sure that revokeObjectURL is called before the next object url is created? I've tried playing with setTimeout and Promises to no avail :(

Reply:
Hi @zaliqarosli `setTimeout` is an asynchronous function and I believe is what's causing your problem. It's also set in the code to 10 milliseconds and will be very fast.